actors/builtin/miner/miner_actor.go

```diff
@@ -217,34 +217,42 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
 	var recoveredSectors []*SectorOnChainInfo
 	penalty := abi.NewTokenAmount(0)
 
+	// Get the total power/reward. We need these to compute penalties.
 	epochReward := requestCurrentEpochBlockReward(rt)
 	pwrTotal := requestCurrentTotalPower(rt)
 
 	rt.State().Transaction(&st, func() interface{} {
 		rt.ValidateImmediateCallerIs(st.Info.Worker)
 
+		// Validate that the miner didn't try to prove too many partitions at once.
 		partitionSize := st.Info.WindowPoStPartitionSectors
 		submissionPartitionLimit := windowPoStMessagePartitionsMax(partitionSize)
 		if uint64(len(params.Partitions)) > submissionPartitionLimit {
 			rt.Abortf(exitcode.ErrIllegalArgument, "too many partitions %d, limit %d", len(params.Partitions), submissionPartitionLimit)
 		}
+
+		// Load and check deadline.
 		currDeadline := st.DeadlineInfo(currEpoch)
 		deadlines, err := st.LoadDeadlines(adt.AsStore(rt))
 		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
 
-		// Traverse earlier submissions and enact detected faults.
-		// This isn't strictly necessary, but keeps the power table up to date eagerly and can force payment
-		// of penalties if locked pledge drops too low.
-		detectedFaultSectors, penalty = detectFaultsThisPeriod(rt, &st, store, currDeadline, deadlines, epochReward, pwrTotal.QualityAdjPower)
-
+		// Double check that the current proving period has started. This should be impossible in practice (XXX: check this?)
```

Comment:
No, this can happen when cron doesn't happen exactly on the final PP epoch, because it's an empty tipset. Or if the cron has borked generally (which is bad, but currently happens when bugs).

Reply:
This is probably well out of our control at this point but have we thought about invoking cron from the validator on every preceding empty tipset before running a tipset's messages?  Haven't thought about how that could go wrong but it feels like a bad coupling of blockchain details with state machine to require logic in the state machine for handling these edge cases.